src/main/java/com/example/petbuddybackend/service/offer/OfferService.java

```diff
@@ -249,4 +264,84 @@ private void editConfigurationSelectedOptions(OfferConfiguration editingConfigur
+                .toList();
+
+        return offerRepository.saveAll(modifiedOffers)
+                .stream()
+                .map(offerMapper::mapToOfferDTO)
+                .toList();
+    }
+
+    private Offer setAvailabilityForOffer(Long offerId, List<AvailabilityRangeDTO> availabilityRanges, String caretakerEmail) {
+
+        Offer offerToModify = getOffer(offerId);
+        assertOfferIsModifyingByOwnerCaretaker(offerToModify, caretakerEmail);
+        Set<Availability> availabilities = createAvailabilities(availabilityRanges, offerToModify);
+
+        if(offerToModify.getAvailabilities() == null) {
+            offerToModify.setAvailabilities(new HashSet<>());
+        }
+        removeAllAvailabilities(offerToModify.getAvailabilities());
+        offerToModify.getAvailabilities().addAll(availabilities);
+
+        return offerToModify;
+    }
+
+    private void removeAllAvailabilities(Set<Availability> availabilities) {
+
+        Set<Availability> copyOfAvailabilities = new HashSet<>(availabilities);
+
+        for(Availability availability : copyOfAvailabilities) {
+            availabilities.remove(availability);
+        }
+
+    }
+
+    private void assertOfferIsModifyingByOwnerCaretaker(Offer offerToModify, String caretakerEmail) {
+        if(!offerToModify.getCaretaker().getEmail().equals(caretakerEmail)) {
+            throw new UnauthorizedException("Caretaker with email: " + caretakerEmail +
+                    "is trying to modify offer that does not belong to him");
+        }
+    }
+
+    private Set<Availability> createAvailabilities(List<AvailabilityRangeDTO> availabilityRanges, Offer offer) {
+        return availabilityRanges.stream()
+                .map(availabilityRange -> createAvailability(availabilityRange, offer))
+                .collect(Collectors.toSet());
+    }
+
+    private Availability createAvailability(AvailabilityRangeDTO availabilityRange, Offer offer) {
+
+        return Availability.builder()
+                .availableFrom(availabilityRange.availableFrom())
+                .availableTo(availabilityRange.availableTo())
+                .offer(offer)
+                .build();
+
+    }
+
+    private void assertAvailabilityRangesNotOverlapping(List<AvailabilityRangeDTO> availabilityRanges) {
+        for(int i = 0; i < availabilityRanges.size(); i++) {
+            for(int j = i + 1; j < availabilityRanges.size(); j++) {
+                if(availabilityRanges.get(i).overlaps(availabilityRanges.get(j))) {
+                    throw new IllegalArgumentException(
+                        MessageFormat.format(
+                            "Availability range {0} overlaps with availability range {1}",
+                            availabilityRanges.get(i).toString(),
+                            availabilityRanges.get(j).toString()
+                        )
+                    );
+                }
+            }
+        }
+    }
```

Comment:
```suggestion
    private void assertAvailabilityRangesNotOverlapping(List<AvailabilityRangeDTO> availabilityRanges) {

        availabilityRanges.sort(Comparator.comparing(AvailabilityRangeDTO::availableFrom));

        for(int i = 1; i < availabilityRanges.size(); i++) {
            AvailabilityRangeDTO previous = availabilityRanges.get(i-1);
            AvailabilityRangeDTO current = availabilityRanges.get(i);
            
            if(previous.overlaps(current)) {
                throw new IllegalArgumentException(
                        MessageFormat.format(
                                "Availability range {0} overlaps with availability range {1}",
                                previous.toString(),
                                current.toString()
                        )
                );
            }
        }
    }
```

To jest w ogóle ciekawy problem i znalazłem, że da się go rozwiązać w O(nlogn):
https://www.geeksforgeeks.org/check-if-any-two-intervals-overlap-among-a-given-set-of-intervals/

Reply:
algorytmiarz